doc/src/explanation/glossary.md

```diff
@@ -0,0 +1,743 @@
+    the top-level {term}`subexpressions <subexpression>` used to create
+    `expr`. They are the arguments to the class used to create the expression.
+    The args of any expression can be obtained by the `.args` attribute. For
+    example, `(1 + x*y).args` is `(1, x*y)`, because it equals `Add(1, x*y)`.
+    The `args` together with {term}`func` completely define an expression. It
+    is always possible to walk the {term}`expression tree` and extract any
+    subexpression of a SymPy expression by repeated use of `.args`. Every
+    SymPy expression can be rebuilt exactly with `func` and `args`, that is,
+    `expr.func(*expr.args) == expr` will always be true of any SymPy
+    expression `expr`. The args of an expression may be the empty tuple `()`,
+    meaning the expression is an {term}`atom`.
+
+Assumptions
+
+    *Assumptions* are a set of predicates on a {term}`symbol` or
+    {term}`expression` that define the set of possible values it can take.
+    Some examples of assumptions are `positive`, `real`, and `integer`.
+    Assumptions are related to one another logically, for example, an
+    assumption of `integer` will always automatically imply `real`.
+    Assumptions use a {term}`three-valued logic` system where predicates are
+    either `True`, `False`, or `None`.
+
+    Assumptions are either *assumed* or *queried*. For example, a symbol `x`
+    might be *assumed* to be positive by defining it as `x = symbols('x',
+    positive=True)`. Then an assumption might be *queried* on the expression
+    containing this symbol, like `(x + 1).is_real`, which in this case would
+    return `True`.
+
+    If no assumptions are assumed on a symbol, then by default symbols are
+    assumed to be general complex numbers. Setting assumptions is important
+    because certain simplifications are only mathematically true in a
+    restricted domain, for example, $\sqrt{x^2} = x$ is not true for general
+    complex $x$ but it is true when $x$ is positive. SymPy functions will
+    never perform an operation on an expression unless it is true for all
+    values allowed by its assumptions.
+
+    SymPy has two separate assumptions systems, which are closely related to
+    one another. The first, which is sometimes called the "old assumptions"
+    because it is older, allows assumptions to be assumed on {term}`Symbol`
+    objects and queried with {term}`is_*` attributes. The second, which is
+    sometimes called the "new assumptions", allows assumptions to be assumed
+    using separate predicate objects like `Q.positive` and does queries using
+    the {func}`~.ask` function. The newer assumptions system is able to
+    support more complex queries, but is also not as well developed as the
+    older one. Most users of SymPy should prefer the older assumptions system
+    at this time.
+
+Atom
+
+    An *atom* is an expression whose {term}`args` is the empty tuple `()`.
```

Comment:
Suggest giving an example of an atom in context. Also, this might be a good place to give the non-intuitive case where, for example, an atom is `x >= 0.3` from `(x >= 0.3) & (x <= 1.0)`.

Reply:
But that isn't an atom:

```
>>> (x >= 0.3).args
(x, 0.300000000000000)
```

If there's some other documentation suggesting this is an atom, that should be fixed, because the definition I wrote here is what we should be using to mean "atom". I rather suspect there may be some instances where documentation elsewhere hasn't been written very rigorously and will need to be updated to match the glossary. We can consider it to be both descriptive and prescriptive.

Reply:
Understand that isn't an atom. However, using the atoms(Relational) method on a solved equation returns that as an "atom":

```
>>> from sympy import *
>>> from sympy.core.relational import Relational
>>> x = symbols('x')
>>> eq = solve([x >= 0.5, x <= 0.7])
>>> [atom for atom in eq.atoms(Relational)]
[0.5 <= x, x <= 0.7]
```

Apparently this is a quirk of using Relational as an argument of atoms() (and perhaps other arguments), because omitting an argument does atoms that meet the definition given:

```
>>> [atom for atom in eq.atoms()]
[x, 0.700000000000000, 0.500000000000000]
```

In the interest of keeping the glossary short, it may well be best to discuss such subtleties elsewhere.

Reply:
Oh, I see. The `atoms` method is a bit confusing relating to the definition I wrote here, because if you pass it a class it just returns all the instances of that class inside of an expression, even if those instances aren't "atomic" expressions. I guess I can add this to the reasons I don't really like the term "atom". I'll need to think about how to clarify this. Maybe a note should just be added to the `atoms` docstring about this.